index.html

```diff
@@ -18,6 +18,9 @@
         <!-- /ko -->
         <button class="btn btn-block btn-danger remove-widget" data-bind="click: $parent.remove">Remove</button>
       </div>
+      <!-- ko if: $index() % 4 === 3 --><div class="clearfix visible-lg-block"></div><!-- /ko -->

```

Comment:
These 3 lines are the magic code that fixes the issue with the widgets not filling up new rows.

See the section titled "Responsive column resets" here:
http://getbootstrap.com/css/#grid-responsive-resets

Also see information about the `$index` observable here under the section titled "Note 2: Using $index, $parent, and other context properties":
http://knockoutjs.com/documentation/foreach-binding.html

Basically, if you look at the example code on Bootstrap's documentation, they have a problem where the first column is taller than the other 3 columns. This is similar to our problem where some of our widgets are taller than others. The requirement is that when you resize the screen (which you should when looking at that example), 2 of the columns should move *below* the other 2. That is the same as our requirement. In that example, they know that the 2 columns are going to move below the other 2 when the screen is set to XS. For SM and above, the columns will never wrap because they are each 3 wide, and that is a total of 12. But for XS, they are each 6 wide, which is a total of 24, so it should occupy 2 rows. Because of this, they have put an empty div with a special class `clearfix` between columns 2 and 3. They have also put a class `visible-xs-block` which means that this special separator div should only be drawn on the screen when the screen is XS.

Now that's very simple and works fine for them. But our scenario is a little more complex. In our scenario, the user can add/remove widgets, so the number of columns can grow and shrink and is dynamically generated. Also, we have 3 break points, not 1. They only ever cater for XS and SM, so they only ever need the clearfix div visible for XS. But we cater for XS, SM, MD and LG. So we have to draw 3 different types of clearfix divs. One of them is for LG screens (the top line). The second is for MD screens (the second line). The third is for SM screens (the third line). Also, each of these divs should only be rendered in the correct places. The first type of div for LG screens where we display 4 widgets per line should be placed after every 4 widgets (which is why it has `if: $index() % 4 === 3`). The second type of div for MD screens where we display 3 widgets per line should be placed after every 3 widgets (which is why it has `if: $index() % 3 === 2`). And the third type of div for SM screens where we display 2 widgets per line should be placed after every 2 widgets (which is why it has `if: $index() % 2 === 1`).

We don't need any special clearfix divs to be displayed on any XS screens because we only ever show 1 widget per line on XS screens so this problem does not apply to XS screens.